test/junit/scala/collection/mutable/OpenHashMapTest.scala

```diff
@@ -39,4 +49,50 @@ class OpenHashMapTest {
     // TODO assertEquals(0, fieldMirror.get.asInstanceOf[Int])
     assertEquals(0, field.getInt(m))
   }
+
+  /** Test that an [[OpenHashMap]] frees references to a deleted key (SI-9522). */
+  @Test
+  def freesDeletedKey {
+    import scala.language.reflectiveCalls
+
+    class MyClass {
+      override def hashCode() = 42
+    }
+
+    val counter = new GraphVisitor() {
+      private[this] var instanceCount: Int = _
+
+      def countInstances(obj: AnyRef) = {
+        instanceCount = 0
+        val walker = new GraphWalker(obj)
+        walker.addVisitor(this)
+        walker.walk
+        instanceCount
+      }
+
+      override def visit(record: GraphPathRecord) {
+        if (record.klass() == classOf[MyClass])  instanceCount += 1
+      }
+    }
```

Comment:
> why did you also want JOL in the classpath?

During previous profiling work, I spent some time looking for ways to shrink the footprint of heavily used classes in the compiler internals. Having JOL around would let us write tests to avoid accidentally increasing the size of something that had been performance tuned.

As shown here, it is also handy for writing tests about the object graph, retention, etc. Replacing the home brew version `assertNotReachable` with a dependency from Oracle makes sense to me.

`assertNotReachable` was my suggestion in https://github.com/scala/scala/pull/4133#issuecomment-65720629 when discussing how to write a test case for a similar bug.

Here are the [JOL samples](http://hg.openjdk.java.net/code-tools/jol/file/tip/jol-samples/src/main/java/org/openjdk/jol/samples/) that show what else it can do.